### PR TITLE
Timeouts Send/Recv

### PR DESCRIPTION
**Pull request by quasiben**
PR adds optional timeouts for send/recv

cc @pentschev

### CODE REVIEW

File: ucp/public_api.py [hunk 1]
```diff
@@ -297,8 +300,10 @@ def closed(self):
             is smaller than nbytes or read-only.
         nbytes: int, optional
             Number of bytes to receive. Default is the whole buffer.
+        timeout: float, optional
+            Number of seconds to wait for receive to complete before timing out
         """
-        await self._ep.recv(buffer, nbytes=nbytes)
+        await asyncio.wait_for(self._ep.recv(buffer, nbytes=nbytes), timeout)
```
Comment (pentschev) [inline ucp/public_api.py]:
```suggestion
        await asyncio.wait_for(self._ep.recv(buffer, nbytes=nbytes), timeout=timeout)
```

Comment (pentschev) [reply ucp/public_api.py]: For consistency with previous usage of `wait_for`.

File: tests/test_send_recv.py [hunk 2]
```diff
@@ -152,3 +152,30 @@ def handle_exception(loop, context):
...
+    listener = ucp.create_listener(timeout_server)
+    client = await ucp.create_endpoint(ucp.get_address(), listener.port)
+
+    msg = bytearray(10 ** 10)
+
+    with pytest.raises(asyncio.TimeoutError):
+        await client.send(msg, timeout=0.001)
+
+    with pytest.raises(asyncio.TimeoutError):
+        await client.recv(msg, timeout=1)
```
Comment (jakirkham) [inline tests/test_send_recv.py]: Could we add a test when it doesn't raise?

**Comment by pentschev** [comment]
rerun tests

**Comment by pentschev** [comment]
Anybody has ideas of what could be happening here? All three builds hanged on the same test: https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fucx-py%2Fprb%2Fucx-py-gpu-build/detail/ucx-py-gpu-build/1209/pipeline#log-2562, I can't reproduce the hang on a DGX-1.

**Comment by jakirkham** [comment]
rerun tests

**Comment by jakirkham** [comment]
rerun tests

**Comment by jakirkham** [comment]
Hmm...maybe [the failures here]( https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fucx-py%2Fprb%2Fucx-py-gpu-build/detail/ucx-py-gpu-build/1310/pipeline ) are legitimate?

**Comment by pentschev** [comment]
I'm not sure we still need this PR, it was an attempt to solve the issue with closing Dask workers that wouldn't report to the scheduler they've terminated, but this wasn't the issue. With that said, I think we can leave this aside for now, maybe even close the PR and revisit it in the future if needed, I think the issue needs to be solved in distributed as described in https://github.com/dask/distributed/issues/3576.

**Comment by madsbk** [comment]
> I'm not sure we still need this PR, it was an attempt to solve the issue with closing Dask workers that wouldn't report to the scheduler they've terminated, but this wasn't the issue. With that said, I think we can leave this aside for now, maybe even close the PR and revisit it in the future if needed, I think the issue needs to be solved in distributed as described in [dask/distributed#3576](https://github.com/dask/distributed/issues/3576).

Fine with me, users can always use `asyncio.wait_for()` themselves if they need a timeout.

**Comment by quasiben** [comment]
Sorry, I should've closed this PR earlier.  I'll do that now

**Pull request closed without merging**